src/OpenTelemetry/Trace/EnrichmentScope.cs

```diff
@@ -0,0 +1,100 @@
+// limitations under the License.
+// </copyright>
+using System;
+using System.Diagnostics;
+using OpenTelemetry.Context;
+
+namespace OpenTelemetry.Trace
+{
+    /// <summary>
+    /// A class for enriching the data of <see cref="Activity"/> objects.
+    /// </summary>
+    public class EnrichmentScope : IDisposable
+    {
+        private static readonly RuntimeContextSlot<EnrichmentScope> RuntimeContextSlot = RuntimeContext.RegisterSlot<EnrichmentScope>("otel.enrichment_scope");
+
+        private bool disposed;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EnrichmentScope"/> class.
+        /// </summary>
+        /// <param name="enrichmentAction">Callback action for performing <see cref="Activity"/> enrichment.</param>
+        /// <param name="target">The target for the enrichment scope.</param>
+        private EnrichmentScope(Action<Activity> enrichmentAction, EnrichmentScopeTarget target)
+        {
+            this.EnrichmentAction = enrichmentAction ?? throw new ArgumentNullException(nameof(enrichmentAction));
+            this.EnrichmentTarget = target;
+
+            this.Parent = Current;
+            this.Parent.Child = this;
+
+            RuntimeContextSlot.Set(this);
+        }
+
+        /// <summary>
+        /// Gets the current <see cref="EnrichmentScope"/>.
+        /// </summary>
+        public static EnrichmentScope Current => RuntimeContextSlot.Get();
+
+        /// <summary>
+        /// Gets the parent <see cref="EnrichmentScope"/>.
+        /// </summary>
+        public EnrichmentScope Parent { get; internal set; }
+
+        /// <summary>
+        /// Gets the child <see cref="EnrichmentScope"/>.
+        /// </summary>
+        public EnrichmentScope Child { get; internal set; }
+
+        /// <summary>
+        /// Gets the enrichment action.
+        /// </summary>
+        public Action<Activity> EnrichmentAction { get; private set; }
+
+        /// <summary>
+        /// Gets the enrichment target.
+        /// </summary>
+        public EnrichmentScopeTarget EnrichmentTarget { get; private set; }
+
+        /// <summary>
+        /// Registers an action that will be called to enrich the next <see cref="Activity"/> processed under the current scope if it has been sampled.
+        /// </summary>
+        /// <param name="enrichmentAction">Action to be called.</param>
+        /// <param name="target">The target for the enrichment scope.</param>
+        /// <returns><see cref="IDisposable"/> to cancel the enrichment scope.</returns>
+        public static IDisposable Begin(Action<Activity> enrichmentAction, EnrichmentScopeTarget target = EnrichmentScopeTarget.NextActivity)
+            => new EnrichmentScope(enrichmentAction, target);
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            if (!this.disposed)
+            {
+                this.EnrichmentAction = null;
+
+                this.Parent.Child = this.Child;
```

Comment:
For safety, do we want to check:
```csharp
if (RuntimeContextSlot.Get() != this)
{
    return; // do nothing
}
```

Reply:
No need to address in this PR, I have these in my TODO list:
* What would happen (worst case) if you start and end activities in different order.
* What would happen (worst case) if you end activity in a different thread.

Reply:
Check out what I just pushed. Tried to make it a bit more defensive and also have it "snap back" to the correct parent should things go out of order.

Reply:
LGTM.
If we move a step further, it'll be HAMT in [PEP0550](https://www.python.org/dev/peps/pep-0550/#implementation).